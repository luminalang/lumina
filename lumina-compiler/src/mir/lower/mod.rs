use super::func::InstInfo;
use super::tyfmt::TyFmtState;
use super::{Current, LangItems, ReadOnlyBytes};
use crate::prelude::*;
use crate::Target;
use ast::NFunc;
use std::ops::Not;

mod expr;
pub use expr::Expr;
pub mod pat;

use derive_new::new;
use lumina_typesystem::{
    Constraint, Forall, GenericMapper, IType, Inference, Static, TEnv, Transformer, Type, Upgrade,
    Var,
};
use lumina_util::Highlighting;
use std::fmt;

#[derive(new)]
pub struct Function {
    pub typing: ConcreteTyping,
    pub lambdas: Map<key::Lambda, Lambda>,
    pub lcaptures: Map<key::Lambda, Vec<key::Bind>>,
    #[new(default)]
    pub no_mangle: bool,
    pub expr: Expr,
}

#[derive(Debug, Clone)]
pub struct ConcreteTyping {
    pub forall: Forall<'static, Static>,
    pub params: Vec<Type>,
    pub returns: Type,
}

impl ConcreteTyping {
    pub fn from_hir<T>(
        forall: Forall<'static, Static>,
        hir: &hir::Typing<T>,
        mut f: impl FnMut(Tr<&T>) -> Type,
    ) -> Self {
        Self {
            forall,
            params: hir.params.values().map(|ty| f(ty.as_ref())).collect(),
            returns: f(hir.returns.as_ref()),
        }
    }
}

#[derive(Debug, Clone, new)]
pub struct CallTypes {
    pub params: Map<key::Param, Type>,
    pub ret: Type,
}

#[derive(new, Debug)]
pub struct Lambda {
    pub typing: ConcreteTyping,
    pub expr: Expr,
}

#[derive(Clone, Debug)]
pub enum Callable {
    Func(M<NFunc>, GenericMapper<Static>),
    Lambda(key::Lambda, GenericMapper<Static>),
    Binding(key::Bind),
    Param(key::Param),
}

#[derive(new)]
pub struct Lower<'a, 's> {
    env: &'a TEnv<'s>,

    read_only_table: &'a mut MMap<key::ReadOnly, (ReadOnlyBytes, Type)>,

    pub current: &'a mut Current,

    items: LangItems,
    fnames: &'a MMap<key::Record, Map<key::Field, Tr<&'s str>>>,
    ftypes: &'a MMap<key::Record, Map<key::Field, Tr<Type>>>,
    vtypes: &'a MMap<key::Sum, Map<key::Variant, Vec<Tr<Type>>>>,

    target: Target,

    #[new(default)]
    pub errors: Vec<FinError>,
    #[new(default)]
    pub trait_object_cast_checks: Vec<(Tr<Type>, Constraint<Static>)>,
    // #[new(default)]
    // lambdas: Map<key::Lambda, Lambda>,
}

pub enum FinError {
    UnreachablePattern(Span),
    MissingPatterns(Span, Vec<pat::MissingPattern>),
    InvalidCast(Tr<Type>, Type),
    LargeCharLiteral(Span),
    BadArrayCount { got: Tr<usize>, exp: Tr<u64> },
    BadGenericArrayCount { got: Tr<usize> },
    DuplicateField(Tr<String>, Span),
    MissingField(String, Span),
}

impl<'a, 's> Lower<'a, 's> {
    pub(super) fn finalizer(&mut self) -> Upgrade {
        Upgrade(self.env)
    }

    fn fin_typing(&mut self, info: &InstInfo) -> (GenericMapper<Static>, CallTypes) {
        let inst = self.fin_inst(&info.inst);
        let mut fin = self.finalizer();
        let ptypes = info.ptypes.iter().map(|ty| fin.transform(&**ty)).collect();
        let returns = fin.transform(&info.ret);
        let call = CallTypes::new(ptypes, returns);
        (inst, call)
    }

    fn fin_inst(&mut self, info: &GenericMapper<Inference>) -> GenericMapper<Static> {
        info.map(|ty| self.finalizer().transform(ty))
    }

    pub fn typing(
        &mut self,
        forall: &Forall<'s, Inference>,
        typing: &hir::Typing<IType>,
    ) -> ConcreteTyping {
        let mut fin = self.finalizer();
        ConcreteTyping {
            params: typing
                .params
                .values()
                .map(|ty| fin.transform(&*ty))
                .collect(),
            returns: fin.transform(&typing.returns),
            forall: forall.map(|_, ty| fin.transform(ty), Forall::<()>::name_by_key),
        }
    }

    fn str_to_ro(&mut self, str: &'s str) -> M<key::ReadOnly> {
        let str = escape(str);

        // We set the type to `u8` because when this data is accessed
        // with ReadOnly, it's treated by-reference so it'll become `*u8`
        let ty = Type::u8();

        self.read_only_table.push(
            self.current.fkey.0,
            (mir::ReadOnlyBytes(str.into_boxed_slice()), ty),
        )
    }

    pub fn patterns_and_expr(
        &mut self,
        params: &[Type],
        patterns: &[Tr<hir::Pattern<'s>>],
        expr: Tr<&hir::Expr<'s>>,
    ) -> Expr {
        assert_eq!(params.len(), patterns.len());

        let mut plower = ParamsLower {
            lower: self,
            params,
            patterns,
            tail: expr,
            current_param: 0,
            lowered_tail: None,
        };
        let _ = pat::Merge::generate_tail(&mut plower);

        plower.lowered_tail.unwrap()
    }
}

fn escape(str: &str) -> Vec<u8> {
    let mut buffer = Vec::with_capacity(str.len());
    let mut bytes = str.bytes();

    loop {
        let Some(mut b) = bytes.next() else {
            break buffer;
        };

        match b {
            b'\\' => match bytes.next() {
                Some(b'n') => b = b'\n',
                Some(b'r') => b = b'\r',
                Some(b't') => b = b'\t',
                Some(b'\\') => b = b'\\',
                Some(b'"') => b = b'"',
                Some(b'\'') => b = b'\'',
                Some(b'0') => b = b'\0',
                b => {
                    panic!("ET: invalid escape sequence: {b:?}");
                }
            },
            _ => {}
        }

        buffer.push(b);
    }
}

pub struct ParamsLower<'l, 'a, 's> {
    lower: &'l mut Lower<'a, 's>,
    params: &'l [Type],
    patterns: &'l [Tr<hir::Pattern<'s>>],
    current_param: usize,

    tail: Tr<&'l hir::Expr<'s>>,
    lowered_tail: Option<Expr>,
}

impl<'l, 'a, 's> pat::Merge<'s, key::DecisionTreeTail> for ParamsLower<'l, 'a, 's> {
    fn generate_tail(&mut self) -> key::DecisionTreeTail {
        match self.params.get(self.current_param) {
            None => {
                if self.lowered_tail.is_none() {
                    self.lowered_tail = Some(self.lower.lower_expr(self.tail));
                }

                key::DecisionTreeTail(0)
            }
            Some(ty) => {
                let i = self.current_param;
                self.current_param += 1;

                let (string, maybe, list) = (
                    self.lower.items.pinfo.string,
                    self.lower.items.pinfo.maybe,
                    self.lower.items.list_default,
                );
                let tree = self.first(string, maybe, list, ty, self.patterns[i].as_ref());

                let missing = pat::MissingGeneration::new(pat::Init::new(
                    self.lower.ftypes,
                    self.lower.vtypes,
                ))
                .run(&tree);

                if !missing.is_empty() {
                    self.lower
                        .errors
                        .push(FinError::MissingPatterns(self.patterns[i].span, missing));
                }

                let on = Expr::Yield(Callable::Param(key::Param(i as u32)));

                let mut tails = Map::new();
                match self.lowered_tail.take() {
                    Some(tail) => {
                        tails.push(tail);
                        let pred = Map::from([1]);
                        self.lowered_tail = Some(Expr::Match(Box::new(on), tree, tails, pred));
                    }
                    None => {
                        self.lowered_tail = Some(Expr::Poison);
                    }
                }

                key::DecisionTreeTail(0)
            }
        }
    }

    fn type_of_bind(&mut self, bind: key::Bind) -> Type {
        let ty = self.lower.current.binds[&bind].value.clone();
        self.lower.finalizer().transform(&ty)
    }

    fn err_duplicate_field(&mut self, field: Tr<&'s str>, previous: Span) {
        self.lower.errors.push(FinError::DuplicateField(
            field.map(str::to_string),
            previous,
        ));
    }

    fn str_to_ro(&mut self, str: &'s str) -> M<lumina_key::ReadOnly> {
        self.lower.str_to_ro(str)
    }

    fn extractor_params(&mut self, params: &[Tr<hir::Expr<'s>>]) -> Vec<mir::Expr> {
        params
            .iter()
            .map(|expr| self.lower.lower_expr(expr.as_ref()))
            .collect()
    }

    fn fin_popped_inst(&mut self, span: Span) -> Option<(GenericMapper<Static>, CallTypes)> {
        self.lower
            .current
            .pop_inst(span)
            .map(|instinfo| self.lower.fin_typing(&instinfo))
    }

    fn fin_record(&mut self, rvar: Var) -> Option<(M<key::Record>, Vec<Type>)> {
        self.lower.transform_rvar(rvar)
    }

    fn name_of_field(&self, record: M<key::Record>, field: key::Field) -> &'s str {
        *self.lower.fnames[record][field]
    }

    fn to_init(&self) -> pat::Init {
        pat::Init::new(self.lower.ftypes, self.lower.vtypes)
    }
}

#[derive(new)]
pub struct MatchBranchLower<'l, 'a, 's> {
    lower: &'l mut Lower<'a, 's>,
    tail: Tr<&'l hir::Expr<'s>>,
    tail_key: key::DecisionTreeTail,
    #[new(default)]
    lowered_tail: Option<Expr>,
}

impl<'l, 'a, 's> pat::Merge<'s, key::DecisionTreeTail> for MatchBranchLower<'l, 'a, 's> {
    fn generate_tail(&mut self) -> key::DecisionTreeTail {
        if self.lowered_tail.is_none() {
            self.lowered_tail = Some(self.lower.lower_expr(self.tail));
        }

        self.tail_key
    }

    fn name_of_field(&self, record: M<key::Record>, field: key::Field) -> &'s str {
        *self.lower.fnames[record][field]
    }

    fn type_of_bind(&mut self, bind: key::Bind) -> Type {
        let ty = self.lower.current.binds[&bind].value.clone();
        self.lower.finalizer().transform(&ty)
    }

    fn err_duplicate_field(&mut self, field: Tr<&'s str>, previous: Span) {
        self.lower.errors.push(FinError::DuplicateField(
            field.map(str::to_string),
            previous,
        ));
    }

    fn str_to_ro(&mut self, str: &'s str) -> M<key::ReadOnly> {
        self.lower.str_to_ro(str)
    }

    fn fin_popped_inst(&mut self, span: Span) -> Option<(GenericMapper<Static>, CallTypes)> {
        self.lower
            .current
            .pop_inst(span)
            .map(|instinfo| self.lower.fin_typing(&instinfo))
    }

    fn fin_record(&mut self, rvar: Var) -> Option<(M<key::Record>, Vec<Type>)> {
        self.lower.transform_rvar(rvar)
    }

    fn extractor_params(&mut self, params: &[Tr<hir::Expr<'s>>]) -> Vec<mir::Expr> {
        params
            .iter()
            .map(|expr| self.lower.lower_expr(expr.as_ref()))
            .collect()
    }

    fn to_init(&self) -> pat::Init {
        pat::Init::new(self.lower.ftypes, self.lower.vtypes)
    }
}

pub fn emit_fin_error<'s>(
    sources: &ast::Sources,
    module: key::Module,
    tfmt: TyFmtState<'_, 's>,
    error: FinError,
) {
    match error {
        FinError::LargeCharLiteral(span) => sources
            .error("invalid char literal")
            .m(module)
            .eline(
                span.move_indice(1),
                "char literals may only be used to represent a single character",
            )
            .emit(),
        FinError::DuplicateField(field, previous) => sources
            .error("field assigned twice")
            .m(module)
            .iline(previous, "first assigned here")
            .iline(field.span, "later assigned again here")
            .emit(),
        FinError::UnreachablePattern(span) => sources
            .warning("unreachable pattern")
            .m(module)
            .eline(span, "")
            .emit(),
        FinError::MissingPatterns(span, missing) => {
            let init = sources
                .error("patterns not exhaustive")
                .m(module)
                .eline(span, "")
                .text("missing patterns");

            let name_of_var =
                |sum: M<key::Sum>, var: key::Variant| tfmt.hir.vnames[sum][var].value.to_string();
            let name_of_field = |record: M<key::Record>, field: key::Field| {
                tfmt.hir.fnames[record][field].value.to_string()
            };

            missing
                .into_iter()
                .fold(init, |err, pat| {
                    err.text(format!("  {}", pat.fmt(&name_of_var, &name_of_field)))
                })
                .emit();
        }
        FinError::BadArrayCount { got, exp } => sources
            .error("incorrect array length")
            .m(module)
            .eline(got.span, format!("has {got} elements"))
            .iline(exp.span, format!("expected {exp}"))
            .emit(),
        FinError::BadGenericArrayCount { got } => sources
            .error("incorrect array length")
            .m(module)
            .eline(
                got.span,
                format!("arrays with generic length cannot be constructed with {got} elements",),
            )
            .emit(),
        FinError::InvalidCast(from, to) => {
            sources
                .error("invalid cast")
                .m(module)
                .eline(
                    from.span,
                    format!(
                        "cannot cast a value of type `{}` to `{}`",
                        tfmt.clone().fmt(&*from),
                        tfmt.fmt(to)
                    ),
                )
                .emit();
        }
        // TODO: merge multiple
        FinError::MissingField(field, rspan) => sources
            .error("missing record field")
            .m(module)
            .eline(rspan, format!("field `{field}` is missing"))
            .emit(),
    }
}

impl fmt::Display for Lambda {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{} {} {}\n  {}",
            "as".keyword(),
            &self.typing,
            '='.symbol(),
            self.expr.to_string().lines().format("\n  "),
        )
    }
}

impl fmt::Display for Function {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let constraints = self
            .typing
            .forall
            .generics
            .iter()
            .filter_map(|(generic, gdata)| {
                let cons = &gdata.trait_constraints;
                cons.is_empty().not().then(|| {
                    format!(
                        "{generic} {} {}",
                        "can".keyword(),
                        cons.iter().format(" + ")
                    )
                })
            })
            .collect::<Vec<_>>();

        let w = "when".keyword();

        match constraints.as_slice() {
            [] => Ok(()),
            [con] => writeln!(f, "{w} {con}"),
            many => writeln!(f, "{w}\n  {}", many.iter().format("\n  ")),
        }?;

        write!(
            f,
            "{} {} {}\n  {}",
            "as".keyword(),
            &self.typing,
            '='.symbol(),
            self.expr.to_string().lines().format("\n  ")
        )?;

        if !self.lambdas.is_empty() {
            write!(
                f,
                "\n {}\n  {}",
                "where".keyword(),
                self.lambdas
                    .iter()
                    .map(|(k, v)| {
                        format!("{} {k} {}", "fn".keyword(), v)
                            .to_string()
                            .lines()
                            .format("\n  ")
                            .to_string()
                    })
                    .format("\n  "),
            )
        } else {
            Ok(())
        }
    }
}

impl fmt::Display for ConcreteTyping {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{}{}{} {} {}{}",
            &self.forall,
            '('.symbol(),
            self.params.iter().format(", "),
            "->".symbol(),
            &self.returns,
            ')'.symbol(),
        )
    }
}
