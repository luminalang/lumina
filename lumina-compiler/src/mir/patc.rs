use super::func::InstInfo;
use super::Verify;
use crate::prelude::*;

use lumina_typesystem::{IType, Ty};

impl<'a, 's> Verify<'a, 's> {
    pub fn type_check_pat(&mut self, pat: Tr<&hir::Pattern<'s>>) -> Tr<IType> {
        trace!("checking pattern {}", pat.value);

        match pat.value {
            hir::Pattern::Int(_, nvar) => IType::infer(*nvar),
            hir::Pattern::Char(_) => IType::u8(),
            hir::Pattern::Any => IType::infer(self.vars().var(pat.span)),
            hir::Pattern::Bind(bind, inner) => {
                let ty = self.type_check_pat((&**inner).tr(pat.span));
                self.new_bind_as(*bind, ty.clone());
                ty.value
            }
            hir::Pattern::Constructor(type_, var, params) => {
                let params = params
                    .iter()
                    .map(|p| self.type_check_pat(p.as_ref()))
                    .collect::<Vec<_>>();
                let (finst, ptypes, returns) = self.type_of_variant(pat.span, *type_, *var);
                let instinfo = InstInfo::new(type_.0, finst, ptypes, returns.clone());
                self.type_check_and_emit_application(pat.span, &params, &instinfo.ptypes);
                returns.value
            }
            hir::Pattern::Record(var, ty, fields) => {
                let fvars = fields
                    .iter()
                    .map(|(name, _, _)| self.vars().add_field(*var, *name))
                    .collect::<Vec<_>>();

                if let Some(ty) = ty {
                    self.type_check_and_emit(IType::infer(*var).tr(pat.span).as_ref(), ty.as_ref());
                }

                for ((name, bind, pat), fvar) in fields.iter().zip(fvars) {
                    let ty = self.type_check_pat(pat.as_ref());
                    let fty = IType::infer(fvar).tr(name.span);
                    self.type_check_and_emit(ty.as_ref(), fty.as_ref());
                    self.new_bind_as(*bind, fty);
                }

                IType::infer(*var)
            }
            hir::Pattern::Tuple(elems) => {
                let types = elems
                    .iter()
                    .map(|p| self.type_check_pat(p.as_ref()).value)
                    .collect();

                IType::tuple(types)
            }

            hir::Pattern::Array(elems, len) => {
                let inner = self.type_check_same("array elements", None, elems, |this, elem| {
                    this.type_check_pat(elem.as_ref())
                });

                IType::array(**len, inner.value)
            }
            hir::Pattern::GenericArray(elems, generic) => {
                let inner = self.type_check_same("array elements", None, elems, |this, elem| {
                    this.type_check_pat(elem.as_ref())
                });

                IType::const_array(**generic, inner.value)
            }

            // TODO: since we decided to desugar patterns early on; we can't use the same
            // `type_check_same_type` as with expressions. That's rather unfortunate.
            //
            // Since MIR is split into two steps. We could technically fix this?
            // no reason why we can't just lower it into a tree on the HIR -> MIR step instead.
            hir::Pattern::Cons(pats, inner_var) => {
                let list = self.items.list_default;

                let value = self.type_check_pat(pats[0].as_ref());
                let exp = IType::infer(*inner_var).tr(pat.span);
                self.type_check_and_emit(value.as_ref(), exp.as_ref());

                let ty = self.type_check_pat(pats[1].as_ref());
                let exp = IType::list(list, vec![IType::infer(*inner_var)]).tr(pats[0].span);
                self.type_check_and_emit(ty.as_ref(), exp.as_ref());

                ty.value
            }

            hir::Pattern::Nil(inner) => {
                let list = self.items.list_default;
                IType::list(list, vec![IType::infer(*inner)])
            }
            hir::Pattern::Bool(_) => IType::bool(),
            hir::Pattern::String(spats) => {
                for (i, spat) in spats.iter().enumerate() {
                    let is_last = || i == spats.len() - 1;
                    match spat {
                        hir::StringPattern::Literal(_) => {}
                        hir::StringPattern::Extractor(extractor) => {
                            let params = self.type_check_params(&extractor.params);
                            self.type_check_pass(
                                pat.span,
                                &extractor.call,
                                &extractor.tanot,
                                params,
                            );
                            if let Some(bind) = extractor.bind {
                                let ty = Ty::string(self.items.pinfo.string, vec![]);
                                self.new_bind_as(bind.value, ty.tr(bind.span));
                            }
                        }
                        hir::StringPattern::Wildcard(bind) => {
                            let ty = if is_last() {
                                IType::string(self.items.pinfo.string, vec![])
                            } else {
                                Ty::u8()
                            };
                            self.new_bind_as(bind.value, ty.tr(bind.span))
                        }
                    }
                }

                let record = self.items.pinfo.string;
                IType::string(record, vec![])
            }
            hir::Pattern::Poison => todo!(),
        }
        .tr(pat.span)
    }

    pub fn type_check_pat_params(
        &mut self,
        ptypes: &Map<key::Param, Tr<IType>>,
        pats: &[Tr<hir::Pattern<'s>>],
    ) {
        for (p, exp) in pats.iter().zip(ptypes.values()) {
            let ty = self.type_check_pat(p.as_ref());
            self.type_check_and_emit(ty.as_ref(), exp.as_ref());
        }
    }
}
