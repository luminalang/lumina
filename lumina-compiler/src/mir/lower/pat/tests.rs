use super::*;
use crate::mir::CallTypes;
use insta::assert_snapshot;
use lumina_collections::MapKey;
use lumina_typesystem::{Generic, GenericKind, GenericMapper, Static, Var};

macro_rules! snapshot_tree_and_missing {
    ($lower:ident, $tree:expr) => {{
        println!("{}", $tree);
        let missing = MissingGeneration::new($lower.to_init()).run(&$tree);
        assert_snapshot!(format!(
            "{}\n\nmissing patterns {{\n  {}\n}}",
            $tree,
            missing
                .into_iter()
                .map(|p| p.fmt(&name_of_var, &name_of_field).to_string())
                .format("\n  ")
        ));
    }};
}

struct Lower {
    tail: Tail,

    ftypes: MMap<key::Record, Map<key::Field, Tr<Type>>>,
    vtypes: MMap<key::Sum, Map<key::Variant, Vec<Tr<Type>>>>,
}

const MAYBE: key::Sum = key::Sum(0);
const LIST: key::Sum = key::Sum(1);
const JUST: key::Variant = key::Variant(0);
const NONE: key::Variant = key::Variant(1);
const STRING: key::Record = key::Record(0);

fn m<T>(v: T) -> M<T> {
    M(key::Module(0), v)
}

fn u8() -> Tr<Type> {
    Type::u8().tr(Span::null())
}

impl Lower {
    fn new() -> Self {
        lumina_util::test_logger();

        fn map<K: MapKey, T, const N: usize>(values: [T; N]) -> Map<K, T> {
            Map::from(values)
        }

        let variants = map([map([
            vec![
                Type::Generic(Generic::new(key::Generic(0), GenericKind::Entity)).tr(Span::null()),
            ],
            vec![],
        ])]);
        let records = map([]);

        let mut ftypes = MMap::new();
        ftypes.add_module(0);
        ftypes[key::Module(0)] = records;

        let mut vtypes = MMap::new();
        vtypes.add_module(0);
        vtypes[key::Module(0)] = variants;

        Lower { tail: 0, ftypes, vtypes }
    }
}

fn name_of_var(M(module, sum): M<key::Sum>, var: key::Variant) -> String {
    assert_eq!(module, key::Module(0));

    match (sum, var) {
        (MAYBE, NONE) => "None",
        (MAYBE, JUST) => "Just",
        _ => panic!("unknown sum variant: {sum}:{var}"),
    }
    .to_string()
}
fn name_of_field(record: M<key::Record>, field: key::Field) -> String {
    match (record, field) {
        _ => panic!("unknown record field: {record}.{field}"),
    }
}

type Tail = usize;

impl<'a, 's> Merge<'s, Tail> for Lower {
    fn generate_tail(&mut self) -> Tail {
        self.tail
    }

    fn err_duplicate_field(&mut self, field: Tr<&'s str>, _: Span) {
        panic!("duplicate field: {field}");
    }

    fn name_of_field(&self, _: M<key::Record>, _: key::Field) -> &'s str {
        todo!();
    }

    fn to_init(&self) -> Init<'_> {
        Init::new(&self.ftypes, &self.vtypes)
    }

    fn str_to_ro(&mut self, _: &'s str) -> M<key::ReadOnly> {
        todo!()
    }

    fn fin_popped_inst(&mut self, _: Span) -> Option<(GenericMapper<Static>, CallTypes)> {
        todo!();
    }

    fn fin_record(&mut self, _: Var) -> Option<(M<lumina_key::Record>, Vec<Type>)> {
        todo!();
    }

    fn type_of_bind(&mut self, _: lumina_key::Bind) -> Type {
        todo!()
    }

    fn extractor_params(&mut self, _: &[Tr<hir::Expr<'s>>]) -> Vec<mir::Expr> {
        todo!()
    }
}

impl Lower {
    fn patterns<'s>(&mut self, ty: Type, pats: &[(Tr<hir::Pattern<'s>>, bool)]) -> DecTree<Tail> {
        let mut iter = pats.iter();

        let (pat, _) = iter.next().unwrap();
        let mut tree = self.first(m(STRING), m(MAYBE), m(LIST.into()), &ty, pat.as_ref());

        for (p, expected) in iter {
            self.tail += 1;
            let old = tree.clone();
            let reachable = self.branch(m(STRING), m(MAYBE), m(LIST.into()), &mut tree, p.as_ref());

            if *expected != reachable {
                panic!("{p} had expected reachability of {expected}\nbefore merge:\n{old}\nafter merge:\n{tree}");
            } else {
                println!("ok");
            }
        }

        tree
    }
}

#[test]
fn instant_wildcard() {
    let mut lower = Lower::new();

    let tuple = Type::tuple(vec![u8().value, u8().value]);
    let any = hir::Pattern::Any.tr(Span::null());
    let tree = lower.first(m(STRING), m(MAYBE), m(LIST.into()), &tuple, any.as_ref());
    snapshot_tree_and_missing!(lower, tree);

    let bind = hir::Pattern::Bind(key::Bind(0), Box::new(any.value.clone())).tr(Span::null());
    let tree = lower.first(m(STRING), m(MAYBE), m(LIST.into()), &tuple, bind.as_ref());
    snapshot_tree_and_missing!(lower, tree);
}

fn just<'s>(p: Tr<hir::Pattern<'s>>) -> Tr<hir::Pattern<'s>> {
    var(MAYBE, JUST, [p])
}
fn none<'s>() -> Tr<hir::Pattern<'s>> {
    var(MAYBE, NONE, [])
}

fn cons<'s>(x: Tr<hir::Pattern<'s>>, xs: Tr<hir::Pattern<'s>>) -> Tr<hir::Pattern<'s>> {
    hir::Pattern::Cons(Box::new([x, xs]), lumina_typesystem::Var::from(0)).tr(Span::null())
}
fn nil<'s>() -> Tr<hir::Pattern<'s>> {
    hir::Pattern::Nil(lumina_typesystem::Var::from(0)).tr(Span::null())
}

fn var<'s, const N: usize>(
    key: key::Sum,
    var: key::Variant,
    params: [Tr<hir::Pattern<'s>>; N],
) -> Tr<hir::Pattern<'s>> {
    hir::Pattern::Constructor(m(key), var, params.to_vec()).tr(Span::null())
}

fn tuple<'s, const N: usize>(elems: [Tr<hir::Pattern<'s>>; N]) -> Tr<hir::Pattern<'s>> {
    hir::Pattern::Tuple(elems.to_vec()).tr(Span::null())
}

fn bind<'s>(bind: u32, then: Tr<hir::Pattern<'s>>) -> Tr<hir::Pattern<'s>> {
    hir::Pattern::Bind(key::Bind(bind), Box::new(then.value)).tr(Span::null())
}

fn any() -> Tr<hir::Pattern<'static>> {
    hir::Pattern::Any.tr(Span::null())
}

fn r<'s>(range: std::ops::Range<u8>) -> Tr<hir::Pattern<'s>> {
    use lumina_parser::pat::Bound;

    let start = if range.start == u8::MIN {
        Bound::Excess
    } else {
        Bound::Pos(range.start as u128)
    };

    let end = if range.end == u8::MAX {
        Bound::Excess
    } else {
        Bound::Pos(range.end as u128)
    };

    hir::Pattern::Int([start, end], lumina_typesystem::Var::from(0)).tr(Span::null())
}

fn maybe(ty: Tr<Type>) -> Tr<Type> {
    Type::defined(m(MAYBE), vec![ty.value]).tr(Span::null())
}

fn list(ty: Tr<Type>) -> Tr<Type> {
    Type::list(m::<key::Sum>(LIST.into()), vec![ty.value]).tr(Span::null())
}

fn tuplet<const N: usize>(types: [Tr<Type>; N]) -> Type {
    Type::tuple(types.map(|ty| ty.value).to_vec())
}

#[test]
fn bind_inside_tuple_nested() {
    let mut lower = Lower::new();
    let tree = lower.patterns(
        tuplet([
            u8(),
            tuplet([u8(), u8(), u8()]).tr(Span::null()),
            u8(),
            maybe(u8()),
        ]),
        &[(
            bind(
                0,
                tuple([
                    bind(1, r(0..255)),
                    bind(
                        2,
                        tuple([bind(3, r(0..255)), bind(4, r(0..255)), bind(5, r(0..255))]),
                    ),
                    bind(6, any()),
                    bind(7, any()),
                ]),
            ),
            true,
        )],
    );

    snapshot_tree_and_missing!(lower, tree);
}

#[test]
fn just_x_nothing() {
    let mut lower = Lower::new();
    let tree = lower.patterns(
        maybe(u8()).value,
        &[
            (just(bind(0, any())), true),
            (just(bind(1, any())), false),
            (none(), true),
        ],
    );

    snapshot_tree_and_missing!(lower, tree);
}

#[test]
fn int_matrix() {
    let mut lower = Lower::new();
    let tree = lower.patterns(
        tuplet([u8(), u8(), u8()]),
        &[
            (tuple([r(0..6), r(50..255), r(0..128)]), true),
            (tuple([r(0..6), r(50..200), r(129..255)]), true),
            (tuple([r(2..3), r(50..200), r(0..255)]), false),
            (tuple([r(2..3), r(50..210), r(0..255)]), true),
        ],
    );

    snapshot_tree_and_missing!(lower, tree);
}

#[test]
fn missing_maybes() {
    let mut lower = Lower::new();
    let tree = lower.patterns(
        tuplet([maybe(u8()), maybe(u8())]),
        &[
            (tuple([none(), just(r(0..255))]), true),
            (tuple([just(r(0..255)), just(r(0..255))]), true),
            (tuple([any(), just(r(0..100))]), false),
            (tuple([none(), none()]), true),
            (tuple([none(), any()]), false),
        ],
    );

    snapshot_tree_and_missing!(lower, tree);
}

#[test]
fn list_x_xs() {
    let mut lower = Lower::new();
    let tree = lower.patterns(
        list(u8()).value,
        &[
            (cons(bind(10, any()), bind(20, any())), true),
            (nil(), true),
        ],
    );

    snapshot_tree_and_missing!(lower, tree);
}

#[test]
fn nested_lists() {
    let mut lower = Lower::new();
    let tree = lower.patterns(
        list(list(tuplet([]).tr(Span::null()))).value,
        &[
            (cons(bind(0, any()), bind(1, any())), true),
            (cons(cons(tuple([]), nil()), nil()), false),
            (any(), true),
        ],
    );

    snapshot_tree_and_missing!(lower, tree);
}
