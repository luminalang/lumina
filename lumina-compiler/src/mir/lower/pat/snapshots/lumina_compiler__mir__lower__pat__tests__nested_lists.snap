---
source: lumina-compiler/src/mir/lower/pat/tests.rs
expression: "format!\n(\"{}\\n\\nmissing patterns {{\\n  {}\\n}}\", tree,\nmissing.into_iter().map(| p |\np.fmt(& name_of_var, & name_of_field).to_string()).format(\"\\n  \"))"
---
list variant0
    list variant0
        tuple(len = 0)
          list variant0
              _
                _
                  {
                    table = [PointTable { binds: [(b0, 5), (b1, 6)] }],
                    excess = [[[()]]]
                    tail = 0
                  }
          list variant1
              {
                table = [PointTable { binds: [(b0, 3), (b1, 4)] }],
                excess = [[[()]]]
                tail = 0
              }
    list variant1
        {
          table = [PointTable { binds: [(b0, 1), (b1, 2)] }],
          excess = [[[()]]]
          tail = 0
        }
list variant1
    {
      table = [PointTable { binds: [] }],
      excess = []
      tail = 2
    }

missing patterns {
  
}
